### PR TITLE
Partial evaluation of EVM optimization lemmas

### PR DESCRIPTION
**Pull request by ehildenb**
This is factored out of: https://github.com/runtimeverification/evm-semantics/pull/2551

Inspired by: https://github.com/runtimeverification/evm-semantics/pull/2565. It appears there is some use in partial evaluation of the EVM optimizations.

This should have the correct workstack sizes and such attached to each optimization.

### CODE REVIEW

File: kevm-pyk/src/kevm_pyk/kproj/evm-semantics/optimizations.md [hunk 1]
```diff
@@ -191,7 +191,7 @@ module EVM-OPTIMIZATIONS
       ...
     </kevm>
     requires ( #if USEGAS #then Gverylow < SCHED > <=Gas GAVAIL #else true #fi )
-     andBool ( #sizeWordStack( chop( ( W0 +Int W1 ) ) : WS ) <=Int 1024 )
+     andBool ( #sizeWordStack( WS ) <=Int 1023 )
```
Comment (PetarMax) [inline kevm-pyk/src/kevm_pyk/kproj/evm-semantics/optimizations.md:194]: If `#sizeWordStack( WS ) ==Int 1023`, then `#sizeWordStack( W0 : W1 : WS ) ==Int 1025`, which is malformed. I believe that it should be: 
```suggestion
     andBool ( #sizeWordStack( WS ) <=Int 1022 )
```

File: kevm-pyk/src/kevm_pyk/kproj/evm-semantics/optimizations.md [hunk 2]
```diff
@@ -227,7 +227,7 @@ module EVM-OPTIMIZATIONS
       ...
     </kevm>
     requires ( #if USEGAS #then Gverylow < SCHED > <=Gas GAVAIL #else true #fi )
-     andBool ( #sizeWordStack( chop( ( W0 -Int W1 ) ) : WS ) <=Int 1024 )
+     andBool ( #sizeWordStack( WS ) <=Int 1023 )
```
Comment (PetarMax) [inline kevm-pyk/src/kevm_pyk/kproj/evm-semantics/optimizations.md:230]:
```suggestion
     andBool ( #sizeWordStack( WS ) <=Int 1022 )
```

File: kevm-pyk/src/kevm_pyk/kproj/evm-semantics/optimizations.md [hunk 3]
```diff
@@ -263,7 +263,7 @@ module EVM-OPTIMIZATIONS
       ...
     </kevm>
     requires ( #if USEGAS #then Gverylow < SCHED > <=Gas GAVAIL #else true #fi )
-     andBool ( #sizeWordStack( W0 &Int W1 : WS ) <=Int 1024 )
+     andBool ( #sizeWordStack( WS ) <=Int 1023 )
```
Comment (PetarMax) [inline kevm-pyk/src/kevm_pyk/kproj/evm-semantics/optimizations.md:266]:
```suggestion
     andBool ( #sizeWordStack( WS ) <=Int 1022 )
```

File: kevm-pyk/src/kevm_pyk/kproj/evm-semantics/optimizations.md [hunk 4]
```diff
@@ -299,7 +299,7 @@ module EVM-OPTIMIZATIONS
       ...
     </kevm>
     requires ( #if USEGAS #then Gverylow < SCHED > <=Gas GAVAIL #else true #fi )
-     andBool ( #sizeWordStack( bool2Word( W0 <Int W1 ) : WS ) <=Int 1024 )
+     andBool ( #sizeWordStack( WS ) <=Int 1023 )
```
Comment (PetarMax) [inline kevm-pyk/src/kevm_pyk/kproj/evm-semantics/optimizations.md:302]:
```suggestion
     andBool ( #sizeWordStack( WS ) <=Int 1022 )
```

File: kevm-pyk/src/kevm_pyk/kproj/evm-semantics/optimizations.md [hunk 5]
```diff
@@ -335,7 +335,7 @@ module EVM-OPTIMIZATIONS
       ...
     </kevm>
     requires ( #if USEGAS #then Gverylow < SCHED > <=Gas GAVAIL #else true #fi )
-     andBool ( #sizeWordStack( bool2Word( W1 <Int W0 ) : WS ) <=Int 1024 )
+     andBool ( #sizeWordStack( WS ) <=Int 1023 )
```
Comment (PetarMax) [inline kevm-pyk/src/kevm_pyk/kproj/evm-semantics/optimizations.md:338]:
```suggestion
     andBool ( #sizeWordStack( WS ) <=Int 1022 )
```

File: kevm-pyk/src/kevm_pyk/kproj/evm-semantics/optimizations.md (hunk 1 above)
Comment (ehildenb) [reply kevm-pyk/src/kevm_pyk/kproj/evm-semantics/optimizations.md:194]: That's not the way the rule is summarized. There post-execution stack check (for overflow) checks that _after_ the needed stack elements have been removed, and the new ones are added, that there isn't a stack overflow. It does not check that the stack _going into_ the operation is overflowing.

These summaries were auto-generated by executing each opcode symbolically directly, which is why they are exactly the expressions that showed up. Putting the tighter check on the stack would be a change to the semantics (a change that is still true, as CI tells us, but is more restrictive).

Comment (ehildenb) [reply kevm-pyk/src/kevm_pyk/kproj/evm-semantics/optimizations.md:194]: The point here is to partially evaluate _exactly_ the expressions that are collected during symbolic execution, which is what happens here.

Comment (PetarMax) [reply kevm-pyk/src/kevm_pyk/kproj/evm-semantics/optimizations.md:194]: Very well.

**Review by PetarMax — APPROVED** [review]

**Pull request merged**